### DISCUSSION TITLE
Question - MongoDB/S3 AWS Implementation

### DISCUSSION THREAD

**Discussion by a-r-db** [Lumieducation/H5P-Nodejs-library#1637] [Q&A] [answered]
TL;DR What is the performance increase of implementing public S3 access and MongoDB. Has it ever been calculated? 

[source](https://docs.lumi.education/npm-packages/h5p-mongos3/mongo-s3-content-storage)

To the best of my knowledge the companies using these libraries are implementing production servers using Amazing Web Services.

I am currently using your h5p-client and h5p-server ReactJS project, with several modifications, to implement a development, testing and production server for my employer.

Very soon I will be implementing the backend.

What is the performance difference between hosting everything inside one backend lambda function and hosting each part separately: MongoDB, S3, etc. for the library, content and metadata, respectively?

Also, is there any idea how much performance will be gained if we use public over private S3 endpoints?

I am preparing and collecting information about this problem, but if somebody with more experience on this issue could give some pointers, I would be much obliged.

Also might be serving data to 100 users at first, then as much as 1000 and probably never more than 10000.

**Comment by sr258** [comment, marked as answer]
Do you mean the functionality to serve files directly from S3 as described in the section „ Increasing scalability by getting content files directly from S3“? Or do you mean how the Mongo/S3 storages compare to the file system storages?

The Mongo/S3 storage classes are certainly superior to file system storage in general, especially if you list content. Their main benefit, however is scalability beyond a single instance as the file system storage classes won‘t work (well) across machines or containers. If you want to use the library in a scaled environment, you‘ll have to use Mongo/S3 (and Redis).

If you use public S3 links as described in the docs section above, the performance gain depends a bit on several factors:
- passing files through your app server will always increase latency; if you host everything inside AWS this should be negligible; as there are relatively few files and large that are served this way (if you only serve content files), I don‘t think the effect will be noticeable
- if you have a massive user base with lots of parallel requests you can decrease the load on the app server, which might increase overall performance

To sum it up, I don‘t think it‘s worth using the public S3 endpoint functionality if you don‘t have a very big user base. You always have to keep in mind that serving files directly from the S3 service is charged for by AWS (as far as I know) and that you have to put up something like CloudFront to really have high speed.

We don‘t have measurements on performance gains. The Mongo / S3 content storage adapter is not in large-scale use that we manage ourselves. I wrote it as contractual work and I don‘t know how exactly the customer is using the adapter and how it performs depending on the configuration. I haven‘t heard of any complaints either, though.

If you want to use the Mongo/S3 library storage adapter, I would highly recommend using CloudFront to serve the library files and to change the library files URL accordingly. H5P creates a ton of HTTP requests and they must be as fast as possible. h5p-server has a cache busting mechanism that will allow safe use of a CDN for library files. In general, I think most of the performance issues can be solved by having a super-fast and cached library files storage. It‘s also important to use the cached library storage to improve metadata access times in the library. The performance gains can be more than 1000% compared to uncached access.

I haven‘t used Amazon Lambdas, as we deploy on a self-hosted Kubernetes. I think you should use AWS S3, DocumentDB and their Redis implemention as backend service. It should be fine to put all H5P functionality into a single lambda. H5P wasn‘t designed for a microservice approach, so splitting the core H5P functionality into several services is really hard, as they only have few routes that contain a lot of functionality. (I have a few thoughts on improving this, but I don‘t have time for this at the moment) Your custom functionality to perform CRUD on H5Ps can be separated into their own lambda, but I don‘t think you really have to.

One think to bear in mind is that you might run into issues if users upload massive .h5p files. The library extracts these files to local temporary storage to perform validation and then persist them in the backend storage. (pure in memory validation turned out to be a problem in the past) From what I‘ve seen there‘s only 500 MB of temp storage in lambdas, so this might be a problem.

Are your user numbers concurrent users or total users? If they are total users, I wouldn‘t worry too much about premature performance optimizations.

**Reply by a-r-db** [reply]
This would be total users, concurrently we are expecting probably max 10, 100, and maybe 1000 for each respective user total. We don't expect on ton of traffic right away.

So essentially what you are saying is, I should not worry about performance gains yet, but if needed the library access times are the biggest issue.

Currently we have 18 videos at about ~8MB to ~14MB maximum for interactive videos.
Each video has about 48 interactions.
If we choose 720p the videos will be ~8MB and if we choose 1080p the videos are around ~13MB each.
The rest of the load will be interactive books (which have yet to be designed) and games pulling from a WebGL/unity hosting site.

Essentially the project is growing out from my experience with Amplify CLI.

I am concerned about transfer costs.
I want the backend and frontend to be on the same Amazon network if needed.

Are public endpoints the way to go or some kind of AWS VPC?

I have not really checked if Amazon will charge double the transfer cost if I'm pulling from the backend and serving to the end user both on public domains.

I need to make sure I keep costs at a minimum.

I was going to simply use Cognito user pools and an admin group to keep users from the "CUD" of CRUD
Really only the admins will have access to the full CRUD of the H5P library using Cognito and Amplify's nodejs API.

And yes, thanks so much for taking the time to comment and discuss this.

I really appreciate your advice.

**Reply by sr258** [reply]
As far as I know, data transfer from S3 is free if it goes to another Amazon service within the same region. If you keep the „app server“ (or whatever you call the lambda) in the same region, you should be fine. You‘re still going to pay for the transfer costs from the app server to the user, obviously. But transfer costs are going to be the least of your worries if you deploy in AWS and don‘t have a huge user base or massive file sizes. Half a year ago we calculated the price for hosting a SaaS H5P system in AWS and it certainly wasn‘t on the cheap side... The various services you need add up and the smallest offerings already looked like an overkill. That‘s why we decided to spin up our own Kubernetes cluster on a few VPS and deploy Minio S3, MongoDB
and Redis as backend services ourselves. Of course, we now have to worry about ops a bit more, but the price is about 1/5th of the AWS price.

Authentication and authorization with Cognito should work. Whenever you pass in a IUser object into one of the library‘s functions, you can add arbitrary data that you get from Cognito. You can then authorize requests in the callbacks that you pass into Mongo/S3 content storage object in the constructor.

One thing to bear in mind is that the H5P editor core sends AJAX requests from places which h5p-nodejs-library can‘t control (except if you use service works to intercept and modify them). They don‘t contain any auth headers and you can only authenticate them using cookies. This only affects the core h5p endpoints, not your CRUD operations on whole content objects. 

I haven‘t used Amplify before, so I can‘t really give tips about using it with h5p-nodejs-library.

I think a VPC would be an overkill if you already use other public Amazon service like DocumentDB or S3. But never having used AWS in depth, I‘m not really qualified to answer here.

**Reply by a-r-db** [reply]
Would it be too much to ask about what service provider you went with?
I was looking at Digital Ocean also.
I completely understand, my dev team prefers Amazon for whatever reason.

**Reply by sr258** [reply]
We tried out Scaleway first, but the peformance was bad (noisy neighbours) and the VPS size wasn’t a good fit for us. That’s why we switched to Hetzner and were very happy. We‘re based in Europe, so we looked for providers that aren’t affected by the US CLOUD Act. There weren‘t any cost effective options that provided all the backend services... so we now manage Mongo, Minio and Redis through Helm ourselves. The nodes are managed by Rancher.

AWS is great if you want to get rid of the ops parts and focus completely on development and if you don‘t have to worry about strict privacy regulations. But in the end, dev time is much more expensive than hosting, so it makes sense to go for whatever you and your team are familiar with.

**Reply by a-r-db** [reply]
Well we are going forward with Amazon for now I guess and I have a question.

For example if I built an h5p package (.h5p) on my computer using Lumi standalone and I had enabled S3 buckets/documentDB for the library, temp storage and content storage, would h5p separate the standalone h5p file into the correct storage devices or not?

**Reply by sr258** [reply]
Yes, sure. That's it's job.

**Reply by a-r-db** [reply]
I'm looking at pricing, and I don't even know if it's within the budget to do Redis/Memcaching or DocumentDB. I'm not sure how much my employer is willing to pay for good performance. I think you were right.
Do you recommend Kubernetes over this kind of micro-service setup?
With your knowledge, what service provider would you recommend in North America to build a comparable setup to the one's you've mentioned?
Would docker be a possibility or is it out of the question?
Lastly, can I use the s3 bucket for content storage without using mongodb or was the module built to be put together?

The benefit is that if we switch to Kubernetes I can still bring all my code with me.

**Reply by a-r-db** [reply]
I suppose the main reason I didn't use Kubernetes (AKS) or Docker (ECS) was upfront cost needed to be low for the employer, so we used the free tier services.

**Reply by sr258** [reply]
If you don't plan on running lots of other services for the customer, the number of components isn't so high to justify the overhead of Kubernetes. You'll have:

- MongoDB
- Redis
- S3
- the NodeJS app server

If you don't need high availability you can simply run these services as Docker containers on a single host in a VPS. You can start off with 2-4 cores and scale up if you CPU load or memory usage becomes too high. As NodeJS is single threaded, you'll need multiple processes of the app server and thus always need Redis to manage caches. 

Before we decided to use Kubernetes (mostly because we also run other services like Matomo, Sentry and FusionAuth which become easier to manage on Kubernetes), I experimented with [CapRover](https://caprover.com). It's basically a self-hosted Heroku that allows you to easily run Docker containers and backend services on one (or several) VMs. Maybe it's a good fit for you and your employer. You can start off with one VM and spread load over multiple VMs if it becomes necessary. From what I've heard DigitalOcean is a good VPS service provider.

**Reply by sr258** [reply]
Sidenote: You don't just need Mongo/S3 for performance reasons but also because the FileContentStorage class doesn't do access permission checks. It's only meant for dev purposes and for desktop environments like Lumi.

**Reply by a-r-db** [reply]
I hate to bother you so much, but your information has been very good so far.
I can buy you a couple coffees if you have a developer coffee address if you want too, I'm not sure my employer would be able to contribute to the project financially however, and I am working part time at min wage, but since this project could be a spring board for future work, I'm dedicating most of my time and resources to it.

My MicroService approach has been a huge headache so far, albeit free to attempt.

They wanted me to use free tier tools, and free tier tools I did use...

Sadly the frontend is fine, but the backend is running in a Lambda function which is free but almost completely useless at the moment.

Lambda functions are completely inadequate for this role of content delivery or so it seems.
They're more one-shot operations, not built for this kind of service.

If my supervisor is certain on using free services do you think my task is doomed to failure?

Getting this working has been eating up my time and probably generating unnecessary waste for the overall project.
And just like you said: Memcache/Redis, DocumentDb, and some kind of public content storage is ideal and 1000% required.
I tried doing public storage to speed things up, but it's just awful responsiveness.

I have two ideas I want to suggest to my supervisor.

1. One time Download 
One is that we create our own custom desktop client based on Lumi Education with a login and download feature.

 - Benefit is long term cost, as we will just have a login portal and hosting files.
 - Detriment is building the Desktop client might be harder than a web app.
 - Might end up stuck with desktop only.
 - Maintenance might be worse than a WebApp.

2. On Demand 
We build a small Digital Ocean Kubernetes Cluster or AWS EKS.

 - Benefit is content delivery for mobile devices and any web ready device.
 - Detriment is long term cost might be higher for on-demand.
 - WebApps have to run 24/7

I personally think they're going to prioritize the web app as not everybody in their user base will have a laptop/pc.

Most likely on demand will be the default choice.
Although I wonder from your standpoint, what is the best route for a start-up education platform with the follow information?

- They desire low long term costs
- Upfront costs need to be reasonable
- Hopefully some coding/configuration hours spend can be reduced

- ~18 interactive videos at about 12mb each, ~18 interactive books at probably 2mb each, ~18 WebGL Games with assets that can probably be optimized to about 3-5mb each (but this does no include the size of h5p assets)
 
So if I were to some how convince my employer and supervisor to switch to Digital Ocean to cut costs since they can always switch back to Amazon if Amazon provides a better deal if we will be using Kubernetes (on-demand) or hosting data storage with an authentication front-end (one-time download).

Can I quote you on this to my employer and supervisor?

>The various services you need add up and the smallest offerings already looked like an overkill. That‘s why we decided to spin up our own Kubernetes cluster on a few VPS and deploy Minio S3, MongoDB
and Redis as backend services ourselves. Of course, we now have to worry about ops a bit more, but the price is about 1/5th of the AWS price.

The problem I have is exactly that. Even with all of the Amazon tools available for micro-services the pricing is about the same or more than running a 3 node Kubernetes cluster for production at Digital Ocean with plenty of computing power for what we're doing.

I think bandwidth in we could keep the cost below $100 USD a month with a Digital Ocean KBS Cluster, whereas Amazon not including bandwidth on their lowest tier services will be nearly $100 dollars a month. Bandwidth in it will probably go much higher.

Essentially we're trying to cut costs wherever possible.

Here's a list of the tools you've recommended so far.
I've grouped them by necessity.

Required? (for basic functionality)
[Minio S3](https://min.io/download#/kubernetes) free self-host
[MongoDB](https://www.mongodb.com/kubernetes) free self-host
[Redis](https://redis.io/) free self-host
[FusionAuth](https://fusionauth.io/pricing/) free self-host
[MailChimp](https://mailchimp.com/) (I think they're already paying for this, so I added it.)
[Wordpress](https://kubernetes.io/docs/tutorials/stateful-application/mysql-wordpress-persistent-volume/) free self-hosted
[Nodejs x2]() free self-hosted

Useful?
[Helm](https://helm.sh/https://helm.sh/) free
[Sentry](https://sentry.io/pricing/) paid

Optional?
[Rancher](https://rancher.com/) paid
[Matomo](https://matomo.org/matomo-on-premise/) free self-host

Am I missing anything?

This is the provider I want to recommend to my employer for pricing and bandwidth pricing.
[Digital Ocean Kubernetes](https://www.digitalocean.com/products/kubernetes/) $60 CAD month I think.. or less if we don't need the 3rd node. $60 or $40.

$20/Month per node ($0.027/hr)
2.5 GB RAM usable (4 GB Total) / 2 vCPUs

**Reply by sr258** [reply]
No coffee donation needed, I'm happy to help out. You can quote me on what I write, but of course, there are no guarantees involved.

Can you use AWS for free to run a H5P service? Possibly. DynamoDB could be a decent replacement for MongoDB and Redis. I think Amazon Glacier would functionally be usable for content data, put it would be too slow. So you need the regular S3 service (paid) for decent speed. Lambdas might work, if you split up the routes... But our library wasn't made for this, so it will probably be cumbersome. If you use DynamoDB you have to write your own storage classes.

But seriously, if your employer hires people to develop an application, why doesn't he simply also pay for hosting? Your cost calculation seem correct to me, but I'd recommend 2vCPUs and 8 GB RAM for Kubernetes worker nodes if you plan want to install FusionAuth or Sentry. They are true memory hogs and we currently have 4 nodes of this size to run everything (Sentry consumes most of the resources). We also have another dedicated Rancher VM that manages the other Kubernetes nodes (can be smaller).

About the services, I'd correct it like this:
required:
Minio S3 free self-host (to reduce the number of services it is also possible to use GridFS and store files in Mongo. There's an unmerged pull request for an implementation that we did for a customer)
MongoDB free self-host
Redis free self-host (could also be replaced by MongoDB)
Nodejs free self-hosted

optional:
FusionAuth free self-host (if you don't want to deal with auth and user management yourself; high resource usage)
Wordpress free self-hosted (not needed for our library)
Helm free (Helm 3 is only a CLI tool that you run on your admin machine, if you use Kubernetes, I would absolutely recommend it as you can deploy all the services with just a few clicks or config files)
Sentry free self-host (useful; we self-host it and it's much cheaper that way; relatively high resource usage)
Rancher free self-host (manages our Kubernetes cluster; with it we can auto-deploy new nodes; also a great GUI for Kubernetes; highly recommend it as it simplifies cluster management a lot; if you can get cluster management from your service provider, you don't need it) 
Matomo free self-host (we use it for analytics; pretty good; doesn't use a lot of resources)

**Reply by a-r-db** [reply]
I'm having a hard time finding Hertzner pricing with Hertzner specs.
It's exactly double the price per node of your setup but at least it will be slightly less than EKS or most of the competitors for managed Kubernetes. 

~= $200 cad
![](https://user-images.githubusercontent.com/8563086/128618107-8fdfcc9e-e5a5-463b-a000-581241b4c51e.png)
~= $350 cad
![](https://user-images.githubusercontent.com/8563086/128618108-91b9ebf7-c3c3-486a-88a9-8a988f08cbb1.png)

I'll need to ask DO if the 4 TB transfer stacks to 12 TB transfer when using a cluster or it's just for each node.

Also, I included Wordpress so I could recommend they migrate their home website to the Kubernetes cluster instead of using the web host provider GoDaddy to consolidate costs.

I guess I really need to know if I mirrored your setup to some extent, would the end result require 4 worker nodes and 2 smaller nodes?
That would push things out of budget I think.

I guess I'm hoping to find out where I can cut corners and what the impact of cutting those corners will be?

i want to run everything as it would be as you described to the best of my ability as a production orchestration on my home workstation.

I really just want to show them proof of concept.

I setup a Virtual Machine with 64GB of RAM and 16vCPUs.

It's currently running Vagrant with 6 nested virtualized machines with the specifications I assume I need.
I am building an install script that installs all the required packages for Kubernetes to start working on Ubuntu 20.04
(I want to monitor resource consumption via console once everything is up and running.)
Then I run a command on the master to get it running and use the produced link to connect the worker machines.

(2vCPU + 8GB ram) x 4
(2vCPU + 4GB ram) x 2

I guess I need to know what the ideal configuration is before I get started and what I can do to cut corners or simplify this installation.

Proof of concept is free and only takes my time and some of my local resources.

I hate to waste your time, but I'm running out of ideas on how to get my employer to find the most cost effective and low maintenance solutions. i'm familiar with docker and configurations setups.

Can I ask how many people and man hours it took to make such a setup and if you translated the overall cost for setup and long-term maintenance what each cost might end up being in CAD?

I know I could find cheaper hosting, but it might take more time to configuring the nodes and ask the hosting provider to connect all the machines with a VPC.

It seems like a long shot that I'll be able to do this on my own, since most of my team is unfamiliar with these technologies, but then again I must try to do my best.

I'm not completely sure Rancher is needed as I have a Kubernetes Dashboard built right into Digital Ocean.

![](https://user-images.githubusercontent.com/8563086/128618622-fe15bf47-0a6f-4782-b990-85dc8b6a9356.png)

Also, as a side note, how much redundancy did you consider for your project?

**Reply by sr258** [reply]
If you get the Kubernetes service from DO, you don't need Rancher and having 4 Nodes will work. 3 nodes will also work if you only install Sentry or FusionAuth and not both. If you don't install either, 3 nodes with 2 vCPU and 4 GB RAM will also work fine. Sentry and FusionAuth are the resource hogs and you'll have to consider whether you want to use them if you have to cut costs. All the rest is really resource efficient. What also uses a lot of CPU is monitoring with Prometheus and Grafana.

The redundancy we have is simply having 4 nodes on which we've spread out MongoDB on 3 of those, Minio on all 4 and we have three pods per app service. That way one of our nodes can fail or be taken out for maintenance without service interruption. Having three nodes is also enough for redundancy sake.

I set up Kubernetes myself and if I disregard dead ends on only count in the setup on Hetzner it took me maybe 15 hours to set everything up including the services. I already have the necessary skills about Kubernetes and Rancher from my day job, so I  didn't have to do much learning here. Maintenance currently is pretty low (probably 1-3 a month), but we don't install every update for every service right away. If your employer is more strict, maintenance cost will also be higher. As we do this as part of our open source effort it's difficult to put a price tag on the work. As our service is free and we the way our company earns money by consultant work and tailoring our software to customer's needs, there's no real price tag on ops. If we'd charge ops costs to customers, Amazon AWS would be much more attractive, as the the high personell costs in Germany mean that their service aren't as expensive in comparison. If your team has to put lots of hours into learning about new tech, simply using the more expensive tech from Amazon might be cheaper for your customer, as they save on hours!

Having a full dev setup on your dev machine might be difficult if you don't have massive amounts of RAM on your machine...

**Reply by a-r-db** [reply]
I have 32 vCPU and 128GB ram. 4GB ram per 1 vCPU

It is running Windows as host os with VMWare Workstation.
Nested inside is 1 ubuntu 20.04 with 64GB ram and 16vCPU
I'm running Vagrant with Virtual-box with nested virtualization to test the machines as a cluster.
I've already setup most of the machines for use as a cluster.
I'll installed Rancher to manage them instead of DO for the time being.

It should be plenty to test with room to spare for my two other VMs.
I don't have a nice car. I decided to buy a nice home dev server/workstation. 😄 
I think we need FusionAuth and I really wish Sentry was an option but I don't think that's possible.

With the resources we've mentioned give or take a few nodes/services,
we could easily service over 1000 users with 3-4 nodes and I assume with 4TB transfer.
That's around the same cost as AWS and we forgot to tell them how important Elasticache will be.
I don't think they included it in the price analysis.

What kind of resources do you think I should get with around $160-$180 USD from digital ocean? $(200 - 225) CAD
Then I'll try and build my server to emulate what DO's best offerings would be and what my employer's desire for month to month cost would be. DO is a great offering cause bandwidth is included in the server cost.

From what I gather it's $180 usd or $240 usd depending on if we can have 3 or 4 nodes.
Actual costs translates to $255.36 cad or $340.48 cad including taxes.
I can build something in no time to show my team how it works and what it looks like.
The nodejs apps can be reconfigured to work with these tools rather quickly.
I'll install sentry on the 3rd or 4th node anyways.
Maybe Sentry can be phased out after the main development stage too,
but it would be so useful I don't think there's any alternative.
Also do you think a single Wordpress instance could be tacked onto this or would it hog too many resources?
That way we could cut something out of their budget and transfer the monthly charge to DO.
I'm sure their WP instance is over priced.

Also note, I don't think anybody else on my team knows how to adapt any of these technologies and it's a very small team. i'm probably the only person on the team at this time who could implement this within a sound amount of time.

Thanks,
Austin.

P.S. Looking to expand your business to Canada 😄 ? 120 CAD sounds like a handsome sum of money to me. I'm only a little bit serious of course. That would be excellent in my case! Although I should probably just start job hunting myself. Real world work experience is much more important than the skills I have obtained from school, research and self-teaching.

**Reply by a-r-db** [reply]
Do you mind telling how many users your setup was provisioned for approximately?
Sharing good numbers might help my employer consider this idea.
Also what type of ingress controller do you use or recommend for this project?
Do you think any kubernetes network plugin is ideal?
I'm looking at Flannel and Calico.

**Reply by a-r-db** [reply]
Well I resigned from the project.
We changed direction again.
Sorry to have wasted your time.

**Reply by sr258** [reply]
Alright, good luck!
